content/posts/2021-05-airdrop-anywhere-part-2.md

```diff
@@ -0,0 +1,167 @@
+
+
+[Last time](/posts/2021-04-airdrop-anywhere-part-1) we broke down the problem of implementing AirDrop so that we can support sending and receiving files from devices that do not natively support AirDrop. After some to and fro we landed on an implementation that involves creating a "proxy" running on an Apple device (which supports AWDL natively) or Linux with [OWL](https://owlink.org).
+
+## Pre-requisites
+
+### Project structure
+I've decided to structure all of this as a "Core" project that'll contain the bulk of the AirDrop implementation. This will be able to be plugged into any .NET Core application in order to provide AirDrop services and each application can provide an implementation of an interface in order to handle file sending and receiving. Our project hierarchy will look like this:
+
+```
+|---AirDropAnywhere.Core
+    |
+    |--AirDropAnywhere.Cli
+    |
+    |--AirdropAnywhere.Web
+```
+
+`AirDropAnywhere.Core` will contain all our shared services (e.g. mDNS advertisements, core AirDrop HTTP API) and the means to configure them in a `WebHost`.
+
+`AirDropAnywhere.Cli` will be a CLI application hosting the services and rendered using [Spectre.Console](https://github.com/spectreconsole/spectre.console). It'll typically be used as a way to send / receive for the current machine via the command line.
+
+Individual devices will connect to an application hosted by `AirDropAnywhere.Web`. I've gone back and forth on this a little - I thought about implementing the bits that allow a device to send / receive content via the "proxy" as a GRPC-streaming service and implementing a UI using [MAUI](https://github.com/dotnet/maui). This satisfies the need of being cross-platform, but requires each device to have a copy of the application installed locally. I _could_ use [gRPC-Web](https://grpc.io/docs/platforms/web/basics/) and hook everything into Blazor or some client-side JS but it feels like a fair bit of boilerplate shennanigans that is elegantly solved by SignalR.
+
+So, next, I considered whether to use a Blazor client-side application using SignalR to connect to the "backend". But, I can't see any great incentive to use Blazor for this kind of application - it doesn't have a huge amount of interactivity and there's a great implementation of SignalR that can be consumed from Javascript ¯\_(ツ)_/¯. In the end I decided that I'd implement this as a bog-standard SPA using [Vue.js](https://vuejs.org/) and connecting to the backend using a SignalR-based "API". I've never used Vue.js but have heard good things - so let's give it a try!
+
+### Tools
+To create the project I'm using a combination of [JetBrains Rider](https://www.jetbrains.com/rider/) and [VS Code](https://code.visualstudio.com/) all running on an ancient 2013 Macbook Pro. I'm using Rider because I've become fond of the speed and refactoring wizardry of the JetBrains toolset and VS Code so that I can remote debug on a Raspberry Pi over SSH (more on that later).
```

Comment:
```suggestion
To create the project I'm using a combination of [JetBrains Rider](https://www.jetbrains.com/rider/) and [VS Code](https://code.visualstudio.com/) all running on an ancient 2013 MacBook Pro. I'm using Rider because I've become fond of the speed and refactoring wizardry of the JetBrains toolset and VS Code so that I can remote debug on a Raspberry Pi over SSH (more on that later).
```